### PR TITLE
feat: `utils.json` module / improving `JsonBone`

### LINKED ISSUE #1064: Use custom decoder and encoder in `JsonBone`
I want to store a `datetime` object in a `JsonBone`. But this isn't serializable.
Can we use the `preprocessJsonObject` and `jsonDecodeObjectHook` from the `tasks` module here too?

### PR DESCRIPTION
**Pull request by phorward**
Proposal to fix #1064.

I'm still a little unhappy with this, as it bites with #1000, but after all it's a different use case. Maybe we can bring both together in some way.

### CODE REVIEW

**Review by ArneGudermann — DISMISSED** [review]

File: src/viur/core/utils/json.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        return [_preprocess_json_object(x) for x in obj]
+
+    return obj
+
+
+def dumps(obj: t.Any) -> str:
+    """
+    Wrapper for json.dumps() which converts additional ViUR datatypes.
+    """
+    return json.dumps(_preprocess_json_object(obj))
```
Comment (sveneberth) [inline src/viur/core/utils/json.py:38]:
```suggestion
def dumps(obj: t.Any, *args, **kwargs) -> str:
    """
    Wrapper for json.dumps() which converts additional ViUR datatypes.
    """
    return json.dumps(_preprocess_json_object(obj), *args, **kwargs)
```

**Review by sveneberth — CHANGES_REQUESTED** [review]

**Review by ArneGudermann — APPROVED** [review]

File: src/viur/core/utils/json.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            return {".__bytes__": base64.b64encode(obj).decode("ASCII")}
+        elif isinstance(obj, db.Entity):
+            # TODO: Handle SkeletonInstance as well?
+            return {
+                ".__entity__": self.default(dict(obj)),
+                ".__key__": db.encodeKey(obj.key) if obj.key else None
+            }
+        elif isinstance(obj, dict):
+            return {self.default(key): self.default(value) for key, value in obj.items()}
+        elif isinstance(obj, (list, tuple, set)):
```
Comment (sveneberth) [inline src/viur/core/utils/json.py:29]: Maybe we should memorize the type. That's helpful for task chunks which want to use `set`s.

**Review by sveneberth — CHANGES_REQUESTED** [review]

**Comment by ArneGudermann** [comment]
Can we fix this https://github.com/viur-framework/viur-core/blob/471bee591af7c1620357bf1a7c24f03abb1bfffc/src/viur/core/bones/relational.py#L17-L21 with this PR too.

File: src/viur/core/utils/json.py (hunk 2 above)
Comment (phorward) [reply src/viur/core/utils/json.py]: see 72203f7149146e55065ef032685f35c6f21759b7

**Review by sveneberth — APPROVED** [review]

**Review by ArneGudermann — APPROVED** [review]

**Pull request merged**